### ISSUE TITLE
frugally-deep: severe unexpected performance loss

### ISSUE THREAD

**Issue by tostaess** [Dobiasd/frugally-deep#66]
I am experiencing a severe performance loss when converting a binary image classification model from Python/Keras to C++ using frugally-deep.

The prediction  time in Python is 1.5 to 2 ms, while the prediction time in C++ is 450 ms. Everything is ran on Windows 7, C++ in Visual Studio 2015.

This is how I measure the times ( I have also included a zip-file with these minimal code examples, as well as the image and the actual h5 and json model):

**Python**
``` Python
import os # Run the code only on CPU
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"
import time
import numpy as np
from keras.models import load_model
from keras.preprocessing.image import load_img, img_to_array

img_width = 40;
img_height = 40;
model = load_model('cppmodel.h5')
image = load_img('Good1612.png')
while 1: # While loop so the time can be measured more than once
    time_image = time.perf_counter() # time before image preprocessing
    image_array = img_to_array(image) * (1. / 255)
    image_array = np.reshape(image_array,(1,img_width, img_height, 3))
    time_start = time.perf_counter() # time before actual prediction
    print(model.predict(image_array))
    time_end = time.perf_counter()
    print('prediction time = %fs' %(time_end - time_start))
    print('prediction + image preprocessing time = %fs' %(time_end - time_image))
    input('Press Enter to continue ...')
```

**C++** (adapted from the OpenCV example in the README)
``` C++
int main()
{
	const auto model = fdeep::load_model("fdeep_model.json");
	const cv::Mat image = cv::imread("Good1612.png");
	std::chrono::high_resolution_clock::time_point time_image = std::chrono::high_resolution_clock::now(); // time before image preprocessing
	cv::cvtColor(image, image, cv::COLOR_BGR2RGB);
	assert(image.isContinuous());
	// Use the correct scaling, i.e. low and high.
	std::chrono::high_resolution_clock::time_point time_before_tensor3 = std::chrono::high_resolution_clock::now(); // time before converting input to tensor3
	const auto input = fdeep::tensor3_from_bytes(image.ptr(),
		static_cast<std::size_t>(image.rows),
		static_cast<std::size_t>(image.cols),
		static_cast<std::size_t>(image.channels())
		, 0.0f, 1.0f);
	std::chrono::high_resolution_clock::time_point time_start = std::chrono::high_resolution_clock::now(); // time before actual prediction
	const double result = model.predict_single_output({ input });
	std::chrono::high_resolution_clock::time_point time_end = std::chrono::high_resolution_clock::now();

	std::chrono::duration<double> time_span_pure_predict = std::chrono::duration_cast<std::chrono::duration<double>>(time_end - time_start);
	std::chrono::duration<double> time_span_tensor3_and_predict = std::chrono::duration_cast<std::chrono::duration<double>>(time_end - time_before_tensor3);
	std::chrono::duration<double> time_span_imagep_and_predict = std::chrono::duration_cast<std::chrono::duration<double>>(time_end - time_image);

	std::cout << "pure prediction time:" << time_span_pure_predict.count() << std::endl;
	std::cout << "tensor3 + prediction time" << time_span_tensor3_and_predict.count() << std::endl;
	std::cout << "image preprocessing + prediction time" << time_span_imagep_and_predict.count() << std::endl;
	std::cout << result << std::endl;
	return 0;
}
``` 
When checking the time for the image preprocessing, the converting of the image to tensor3 and the actual prediction as in the minimal example, the prediction itself is the actual bottleneck.

I'm wondering if this is normal behaviour, or if something would be off. From the performances listed in the README, I'm guessing frugally-deep should normally do a lot better and something is wrong with my code or implementation. Do you have any thoughts on this?

[classification_timing.zip](https://github.com/Dobiasd/frugally-deep/files/2149172/classification_timing.zip)

**Comment by tostaess** [comment]
That gives me results in the order of milliseconds as well indeed. Thanks for the fast response and apologies for using your time with this apparantly easy issue!
